hail/python/hail/methods/pca.py

```diff
@@ -207,13 +207,249 @@ def pca(entry_expr, k=10, compute_loadings=False) -> Tuple[List[float], Table, T
     return hl.eval(g.eigenvalues), scores, None if t is None else t.drop('eigenvalues', 'scores')
 
 
-@typecheck(entry_expr=expr_float64,
+class TallSkinnyMatrix:
+    def __init__(self, block_table, block_expr, source_table, col_key):
+        self.col_key = col_key
+        first_block = block_expr.take(1)[0]
+        self.ncols = first_block.shape[1]
+        self.block_table = block_table
+        self.block_expr = block_expr
+        self.source_table = source_table
+
+
+def _make_tsm(entry_expr, block_size):
+    mt = matrix_table_source('_make_tsm/entry_expr', entry_expr)
+    A, ht = mt_to_table_of_ndarray(entry_expr, block_size, return_checkpointed_table_also=True)
+    A = A.persist()
+    return TallSkinnyMatrix(A, A.ndarray, ht, list(mt.col_key))
+
+
+def _make_tsm_from_call(call_expr, block_size, mean_center=False, hwe_normalize=False):
+    mt = matrix_table_source('_make_tsm/entry_expr', call_expr)
+    mt = mt.select_entries(__gt=call_expr.n_alt_alleles())
+    if mean_center or hwe_normalize:
+        mt = mt.annotate_rows(__AC=agg.sum(mt.__gt),
+                              __n_called=agg.count_where(hl.is_defined(mt.__gt)))
+        mt = mt.filter_rows((mt.__AC > 0) & (mt.__AC < 2 * mt.__n_called))
+
+        n_variants = mt.count_rows()
+        if n_variants == 0:
+            raise FatalError("_make_tsm: found 0 variants after filtering out monomorphic sites.")
+        info(f"_make_tsm: found {n_variants} variants after filtering out monomorphic sites.")
+
+        mt = mt.annotate_rows(__mean_gt=mt.__AC / mt.__n_called)
+        mt = mt.unfilter_entries()
+
+        mt = mt.select_entries(__x=hl.or_else(mt.__gt - mt.__mean_gt, 0.0))
+
+        if hwe_normalize:
+            mt = mt.annotate_rows(
+                __hwe_scaled_std_dev=hl.sqrt(mt.__mean_gt * (2 - mt.__mean_gt) * n_variants / 2))
+            mt = mt.select_entries(__x=mt.__x / mt.__hwe_scaled_std_dev)
+    else:
+        mt = mt.select_entries(__x=mt.__gt)
```

Comment:
One of my goals here was to refactor the `hwe_normalize` method into a more usable form. Right now, it's not much good outside of `hwe_normalized_pca`, because `ht.annotate(x=hwe_normalize(ht.GT))` doesn't work (the expression returned by `hwe_normalize` has a different base table). After this PR, `hwe_normalize` is only used in the Spark pca, so it will go away when that does.

Reply:
That makes sense, disregard my earlier comment then.